File: src/chessbot/chessbot3.py
import chess
import chess.svg
import chess.pgn

import numpy as np
import time

from bisect import insort


piece_values = {chess.PAWN: 100, chess.KNIGHT: 320, chess.BISHOP: 330, chess.ROOK: 500, chess.QUEEN: 900, chess.KING: 20000}

PAWN_TABLE = [
    0, 0, 0, 0, 0, 0, 0, 0,
    5, 10, 10, -20, -20, 10, 10, 5,
    5, -5, -10, 0, 0, -10, -5, 5,
    0, 0, 0, 20, 20, 0, 0, 0,
    5, 5, 10, 25, 25, 10, 5, 5,
    10, 10, 20, 30, 30, 20, 10, 10,
    50, 50, 50, 50, 50, 50, 50, 50,
    0, 0, 0, 0, 0, 0, 0, 0
]

KNIGHTS_TABLE = [
    -50, -40, -30, -30, -30, -30, -40, -50,
    -40, -20, 0, 5, 5, 0, -20, -40,
    -30, 5, 10, 15, 15, 10, 5, -30,
    -30, 0, 15, 20, 20, 15, 0, -30,
    -30, 5, 15, 20, 20, 15, 5, -30,
    -30, 0, 10, 15, 15, 10, 0, -30,
    -40, -20, 0, 0, 0, 0, -20, -40,
    -50, -40, -30, -30, -30, -30, -40, -50,
]


BISHOPS_TABLE = [
    -20, -10, -10, -10, -10, -10, -10, -20,
    -10, 5, 0, 0, 0, 0, 5, -10,
    -10, 10, 10, 10, 10, 10, 10, -10,
    -10, 0, 10, 10, 10, 10, 0, -10,
    -10, 5, 5, 10, 10, 5, 5, -10,
    -10, 0, 5, 10, 10, 5, 0, -10,
    -10, 0, 0, 0, 0, 0, 0, -10,
    -20, -10, -10, -10, -10, -10, -10, -20
]


ROOKS_TABLE = [
    0, 0, 0, 5, 5, 0, 0, 0,
    -5, 0, 0, 0, 0, 0, 0, -5,
    -5, 0, 0, 0, 0, 0, 0, -5,
    -5, 0, 0, 0, 0, 0, 0, -5,
    -5, 0, 0, 0, 0, 0, 0, -5,
    -5, 0, 0, 0, 0, 0, 0, -5,
    5, 10, 10, 10, 10, 10, 10, 5,
    0, 0, 0, 0, 0, 0, 0, 0
]

QUEENS_TABLE = [
    -20, -10, -10, -5, -5, -10, -10, -20,
    -10, 0, 0, 0, 0, 0, 0, -10,
    -10, 0, 5, 5, 5, 5, 0, -10,
    -5, 0, 5, 5, 5, 5, 0, -5,
    0, 0, 5, 5, 5, 5, 0, -5,
    -10, 5, 5, 5, 5, 5, 0, -10,
    -10, 0, 5, 0, 0, 0, 0, -10,
    -20, -10, -10, -5, -5, -10, -10, -20
]


KINGS_TABLE_OPENING = [
    20, 30, 10,  0,  0, 10, 30, 20,
    20, 20,  0,  0,  0,  0, 20, 20,
    -10,-20,-20,-20,-20,-20,-20,-10,
    -20,-30,-30,-40,-40,-30,-30,-20,
    -30,-40,-40,-50,-50,-40,-40,-30,
    -30,-40,-40,-50,-50,-40,-40,-30,
    -30,-40,-40,-50,-50,-40,-40,-30,
    -30,-40,-40,-50,-50,-40,-40,-30
]


KINGS_TABLE_ENDGAME = [
    -50,-40,-30,-20,-20,-30,-40,-50,
    -30,-20,-10,  0,  0,-10,-20,-30,
    -30,-10, 20, 30, 30, 20,-10,-30,
    -30,-10, 30, 40, 40, 30,-10,-30,
    -30,-10, 30, 40, 40, 30,-10,-30,
    -30,-10, 20, 30, 30, 20,-10,-30,
    -30,-30,  0,  0,  0,  0,-30,-30,
    -50,-30,-30,-30,-30,-30,-30,-50
]



def preprocess_openings(file_path):
    fen_to_best_move = {}
    with open(file_path, 'r') as file:
        for line in file:
            parts = line.strip().split(" | Score: ")
            moves_str, score_str = parts[0], parts[1]
            moves = moves_str.split()
            score = int(score_str)

            board = chess.Board()
            for move_san in moves:
                move = board.parse_san(move_san)
                
                # Generate FEN for the current position
                fen = board.fen()
                if fen not in fen_to_best_move:
                    fen_to_best_move[fen] = [(score, move)]
                else:
                    insort(fen_to_best_move[fen], (score, move), key=lambda x: x[0])
                board.push(move)

    return fen_to_best_move

class ChessBot3:
    def __init__(self, depth=2, time_limit=10):
        self.depth = depth
        self.time_limit = time_limit
        self.start_time = 0
        
        self.history_table = np.zeros((64, 64), dtype=np.int32)  # History scores for each possible move
        
        self.fen_to_best_move = preprocess_openings('src/chessbot/openings_filtered.pgn')
        self.is_opening = True

    def update_history_score(self, move, depth):
        self.history_table[move.from_square][move.to_square] += depth ** 2

    def get_move_history_score(self, move):
        return self.history_table[move.from_square][move.to_square]
    
    def get_best_move(self, board):
        if self.is_opening:
            if board.fen() in self.fen_to_best_move:
                if board.turn == chess.WHITE:
                    best_move = self.fen_to_best_move[board.fen()][-1][1]  # Try randomly top 5 moves also
                else:
                    best_move = self.fen_to_best_move[board.fen()][0][1]
                return best_move
            self.is_opening = False

        # TODO: Make it smarter than just this
        if self.is_endgame(board):
            best_move = self.select_move(board, depth=self.depth+1, is_maximizing=board.turn == chess.WHITE)
        else:
            best_move = self.select_move(board, depth=self.depth, is_maximizing=board.turn == chess.WHITE)
        return best_move

    def select_move(self, board, depth=2, alpha=float('-inf'), beta=float('inf'), is_maximizing=True):
        legal_moves = list(board.legal_moves)
        legal_moves.sort(key=self.get_move_history_score, reverse=True)

        best_eval = float('-inf') if is_maximizing else float('inf')
        best_move = None
        for move in legal_moves:
            board.push(move)
            eval = self.__alpha_beta_minimax_helper(board, depth - 1, alpha, beta, not is_maximizing)
            board.pop()

            if is_maximizing and eval > best_eval:
                best_eval = eval
                best_move = move
            if not is_maximizing and eval < best_eval:
                best_eval = eval
                best_move = move

            if is_maximizing: alpha = max(alpha, eval)
            else: beta = min(beta, eval)
            if beta <= alpha:  
                    self.update_history_score(move, self.depth - depth)
                    break
        return best_move

    def __alpha_beta_minimax_helper(self, board, depth=3, alpha=float('-inf'), beta=float('inf'), is_maximizing=True):
        if depth == 0 or board.is_game_over():
            return self.evaluate_board(board, depth)
        
        legal_moves = list(board.legal_moves)
        legal_moves.sort(key=self.get_move_history_score, reverse=True)

        best_eval = float('-inf') if is_maximizing else float('inf')
        for move in legal_moves:
            board.push(move)
            eval = self.__alpha_beta_minimax_helper(board, depth - 1, alpha, beta, not is_maximizing)
            board.pop()

            best_eval = max(best_eval, eval) if is_maximizing else min(best_eval, eval)

            if is_maximizing: alpha = max(alpha, eval)
            else: beta = min(beta, eval)
            if beta <= alpha:
                self.update_history_score(move, self.depth - depth)
                break
        return best_eval

    def evaluate_board(self, board, depth): 
        global piece_values

        if board.is_checkmate():
            if board.turn == chess.WHITE:
                return -1e6 - depth
            else:
                return 1e6 + depth
        if board.is_stalemate() or board.is_insufficient_material() or board.is_fifty_moves() or board.is_repetition(3):
            return 0

        # Material values
        piece_square_tables = {
            chess.PAWN: PAWN_TABLE,
            chess.KNIGHT: KNIGHTS_TABLE,
            chess.BISHOP: BISHOPS_TABLE,
            chess.ROOK: ROOKS_TABLE,
            chess.QUEEN: QUEENS_TABLE,
            chess.KING: KINGS_TABLE_OPENING if not self.is_endgame(board) else KINGS_TABLE_ENDGAME,
        }

        boardvalue = 0
        for piece_type in piece_values.keys():
            for color in [chess.WHITE, chess.BLACK]:
                sign = 1 if color == chess.WHITE else -1
                pieces = board.pieces(piece_type, color)
                material_value = piece_values[piece_type] * len(pieces)
                positional_value = sum(piece_square_tables[piece_type][i] for i in pieces)
                if color == chess.WHITE:
                    positional_value = sum(piece_square_tables[piece_type][i] for i in pieces)
                else:  # Correctly mirror and apply positional values for black
                    positional_value = sum(piece_square_tables[piece_type][chess.square_mirror(i)] for i in pieces)
                boardvalue += sign * (material_value + positional_value)

        return boardvalue
    
    def is_endgame(self, board):
        has_queens = chess.QUEEN in board.piece_map().values()
        num_pieces = len(board.piece_map()) 
        return (not has_queens and num_pieces < 15) or num_pieces < 12
    
    # TODO: Endgames aren't working properly => Increase depth
    # TODO: It didn't take a free rook for some reason, it also didn't see that the rook was for hanging (there might be a small mistake somewhere)
    # TODO: Add previous calcs. memoization
    # TODO: After adding memoization, add iterative deepening

File: src/chessbot_cpp/chessbot3.cpp
#include <iostream>

#include <unordered_map>
#include <string>
#include <fstream>

#include "chessbot3.h"
#include <climits>


#define MIN(a, b) ((a) < (b) ? (a) : (b))
#define MAX(a, b) ((a) > (b) ? (a) : (b))




struct ScoreMovePair {
    int score;
    chess::Move move; // Assuming Move is a suitable type for your chess library

    ScoreMovePair(int score, chess::Move move) : score(score), move(move) {}
};

// Comparator for sorting ScoreMovePair by score
bool compareScoreMovePairs(const ScoreMovePair& a, const ScoreMovePair& b) {
    return a.score < b.score;
}

std::map<std::string, std::vector<ScoreMovePair>> preprocessOpenings(const std::string& file_path) {
    std::map<std::string, std::vector<ScoreMovePair>> fenToBestMove;
    std::ifstream file(file_path);
    std::string line;

    if (!file.is_open()) {
        std::cerr << "Failed to open file: " << file_path << std::endl;
        return fenToBestMove; // Return empty map if file can't be opened
    }

    while (getline(file, line)) {
        auto delimiterPos = line.find(" | Score: ");
        std::string movesStr = line.substr(0, delimiterPos);
        int score = std::stoi(line.substr(delimiterPos + 10)); // "+ 10" to skip past " | Score: "

        chess::Board board; // Assuming Board is a suitable type for your chess library
        std::istringstream movesStream(movesStr);
        std::string moveSan;
        while (movesStream >> moveSan) {
            chess::Move move = chess::uci::parseSan(board, moveSan); // Assuming parseSan is a method to parse SAN move notation
            board.makeMove(move); // Assuming makeMove applies a move to the board

            std::string fen = board.getFen(); // Assuming toFen generates a FEN string for the board state
            ScoreMovePair pair(score, move);
            auto& movesList = fenToBestMove[fen];
            if (movesList.empty() || std::find_if(movesList.begin(), movesList.end(), [&pair](const ScoreMovePair& existingPair) {
                return existingPair.move == pair.move; // Assuming Move has an equality operator
            }) == movesList.end()) {
                movesList.push_back(pair);
            }

            // This might need adjustment depending on how your chess library handles move equality and sorting
        }

        // Sorting each vector of moves by score after all moves are inserted
        for (auto& [fen, pairs] : fenToBestMove) {
            std::sort(pairs.begin(), pairs.end(), compareScoreMovePairs);
        }
    }

    file.close();
    return fenToBestMove;
}

class ChessBot3 {
private:
    int depth;
    int time_limit;
    int start_time;
    std::vector<std::vector<int>> history_table;
    std::map<std::string, std::vector<ScoreMovePair>> fen_to_best_move;
    bool is_opening;

public:
    ChessBot3(int depth = 2, int time_limit = 10) : depth(depth), time_limit(time_limit), start_time(0), is_opening(true) {
        history_table.resize(64, std::vector<int>(64, 0));
        fen_to_best_move = preprocessOpenings("src/chessbot/openings_filtered.pgn");
    }

    void update_history_score(chess::Move move, int depth) {
        history_table[move.from().index()][move.to().index()] += depth * depth;
    }

    int get_move_history_score(chess::Move move) {
        return history_table[move.from().index()][move.to().index()];
    }

    chess::Move get_best_move(chess::Board board) {
        if (is_opening) {
            if (fen_to_best_move.count(board.getFen())) {
                if (board.sideToMove() == chess::Color::WHITE) {
                    return fen_to_best_move[board.getFen()].back().move;
                } else {
                    return fen_to_best_move[board.getFen()].front().move;
                }
            }
            is_opening = false;
        }

        chess::Move best_move;
        if (is_endgame(board)) {
            best_move = select_move(board, depth + 1, board.sideToMove() == chess::Color::WHITE);
        } else {
            best_move = select_move(board, depth, board.sideToMove() == chess::Color::WHITE);
        }
        return best_move;
    }

    chess::Move select_move(chess::Board board, int depth = 2, float alpha=INT_MIN, float beta=INT_MAX, bool is_maximizing = true) {
        chess::Movelist legal_moves;
        chess::movegen::legalmoves(legal_moves, board);

        std::sort(legal_moves.begin(), legal_moves.end(), [this](const chess::Move& a, const chess::Move& b) {
            return get_move_history_score(a) > get_move_history_score(b);
        });

        int best_eval = is_maximizing ? INT_MIN : INT_MAX;
        chess::Move best_move;
        for (const auto& move : legal_moves) {
            board.makeMove(move);
            int eval = alpha_beta_minimax_helper(board, depth - 1, alpha, beta, not is_maximizing);
            board.unmakeMove(move);

            if ((is_maximizing && eval > best_eval) || (!is_maximizing && eval < best_eval)) {
                best_eval = eval;
                best_move = move;
            }

            if (is_maximizing) {
                alpha = MAX(alpha, eval);
            } else {
                beta = MIN(beta, eval);
            }

            if (beta <= alpha) {
                update_history_score(move, depth - this->depth);
                break;
            }
        }
        return best_move;
    }

    int alpha_beta_minimax_helper(chess::Board board, int depth = 3, int alpha = INT_MIN, int beta = INT_MAX, bool is_maximizing = true) {
        if (depth == 0 or board.isGameOver().first != chess::GameResultReason::NONE) {
            return evaluate_board(board, depth);
        }

        chess::Movelist legal_moves;
        chess::movegen::legalmoves(legal_moves, board);
        std::sort(legal_moves.begin(), legal_moves.end(), [this](const chess::Move& a, const chess::Move& b) {
            return get_move_history_score(a) > get_move_history_score(b);
        });

        int best_eval = is_maximizing ? INT_MIN : INT_MAX;
        for (const auto& move : legal_moves) {
            board.makeMove(move);
            int eval = alpha_beta_minimax_helper(board, depth - 1, alpha, beta, !is_maximizing);
            board.unmakeMove(move);

            if (is_maximizing) {
                best_eval = std::max(best_eval, eval);
                alpha = std::max(alpha, eval);
            } else {
                best_eval = std::min(best_eval, eval);
                beta = std::min(beta, eval);
            }

            if (beta <= alpha) {
                update_history_score(move, depth - this->depth);
                break;
            }
        }
        return best_eval;
    }

    int evaluate_board(chess::Board board, int depth) {
        auto temp = board.isGameOver();
        chess::GameResultReason grr = temp.first;
        chess::GameResult gres = temp.second;
        if (grr == chess::GameResultReason::CHECKMATE) {
            if (board.sideToMove() == chess::Color::WHITE) return -1000000;  // TODO: Check if this is correct
            else return 1000000;
        }
        std::map<chess::PieceType, const std::vector<std::vector<int>>> piece_square_tables = {
            {chess::PieceType::PAWN, PAWN_TABLE},
            {chess::PieceType::KNIGHT, KNIGHTS_TABLE},
            {chess::PieceType::ROOK, ROOKS_TABLE},
            {chess::PieceType::BISHOP, BISHOPS_TABLE},
            {chess::PieceType::QUEEN, QUEENS_TABLE},
        };

        if (is_endgame(board)) {
            piece_square_tables.insert({chess::PieceType::KING, KINGS_TABLE_OPENING});
        } else {
            piece_square_tables.insert({chess::PieceType::KING, KINGS_TABLE_ENDGAME});
        }

        int board_value = 0;
        for (const auto& [piece_type, value] : PIECE_VALUES) {
            for (const auto& color : {chess::Color::WHITE, chess::Color::BLACK}) {
                int sign = color == chess::Color::WHITE ? 1 : -1;
                auto pieces = board.pieces(piece_type, color);
                int material_value = value * pieces.size();
                int positional_value = 0;
                for (const auto& square : pieces) {
                    if (color == chess::Color::WHITE) {
                        positional_value += piece_square_tables[piece_type][square];
                    } else {
                        positional_value += piece_square_tables[piece_type][chess::squareMirror(square)];
                    }
                }
                board_value += sign * (material_value + positional_value);
            }
        }
        return board_value;
    }

    bool is_endgame(chess::Board board) {
        bool has_queens = false;
        for (const auto& [square, piece] : board.getPieceMap()) {
            if (piece.getType() == chess::QUEEN) {
                has_queens = true;
                break;
            }
        }
        int num_pieces = board.getPieceMap().size();
        return (!has_queens && num_pieces < 15) || num_pieces < 12;
    }
};

